internal/pkg/agent/storage/store/migrations.go

```diff
@@ -0,0 +1,204 @@
+	}
+
+	if action.Type != fleetapi.ActionTypePolicyChange {
+		log.Warnf("unexpected action type when migrating from action store. "+
+			"Found %s, but only %s is suported. Ignoring action and proceeding.",
+			action.Type, fleetapi.ActionTypePolicyChange)
+		// If it isn't ignored, the agent will be stuck here and require manual
+		// intervention to fix the store.
+		return nil
+	}
+
+	stateStore, err := NewStateStore(log, stateDiskStore)
+	if err != nil {
+		return err
+	}
+
+	// set actions from the action store to the state store
+	stateStore.SetAction(&fleetapi.ActionPolicyChange{
+		ActionID:   action.ActionID,
+		ActionType: action.Type,
+		Data: fleetapi.ActionPolicyChangeData{
+			Policy: conv.YAMLMapToJSONMap(action.Policy),
+		},
+	})
+
+	err = stateStore.Save()
+	if err != nil {
+		log.Debugf("failed to save agent state store, err: %v", err)
+	}
+	return err
+}
+
+// migrateYAMLStateStoreToStateStoreV1 migrates the YAML store to the new JSON
+// state store. If the contents of store is already a JSON, it returns the
+// parsed JSON.
+func migrateYAMLStateStoreToStateStoreV1(log *logger.Logger, store storage.Storage) error {
+	exists, err := store.Exists()
+	if err != nil {
+		return fmt.Errorf("migration YMAL to Store v1 failed: "+
+			"could not load store from disk: %w", err)
+	}
+
+	// nothing to migrate, return empty store.
+	if !exists {
+		return nil
+	}
+
+	// JSON is a subset of YAML, so first check if it's already a JSON store.
+	reader, err := store.Load()
+	if err != nil {
+		return fmt.Errorf("could not read store content: %w", err)
+	}
+
+	st, err := readState(reader)
+	// close it as soon as possible and before the next store save
+	_ = reader.Close()
+	if err == nil {
+		// it's a valid JSON, therefore nothing to migrate
+		return nil
+	}
+
+	// Try to read the store as YAML
+	yamlStore, err := migrations.LoadYAMLStateStore(store)
+	if err != nil {
+		// it isn't a YAML store
+		return errors.Join(ErrInvalidYAML, err)
+	}
+
+	// Store was empty, nothing to migrate
+	if yamlStore == nil {
+		return nil
+	}
+
+	var action fleetapi.Action
+	switch yamlStore.Action.Type {
+	case fleetapi.ActionTypePolicyChange:
+		action = &fleetapi.ActionPolicyChange{
+			ActionID:   yamlStore.Action.ActionID,
+			ActionType: yamlStore.Action.Type,
+			Data: fleetapi.ActionPolicyChangeData{
+				Policy: conv.YAMLMapToJSONMap(yamlStore.Action.Policy)},
+		}
+	case fleetapi.ActionTypeUnenroll:
+		action = &fleetapi.ActionUnenroll{
+			ActionID:   yamlStore.Action.ActionID,
+			ActionType: yamlStore.Action.Type,
+			IsDetected: yamlStore.Action.IsDetected,
+		}
+	default:
+		log.Warnf("loaded a unsupported %s action from the deprecated YAML state store, ignoring it",
+			yamlStore.Action.Type)
+	}
+
+	var queue actionQueue
+	for _, a := range yamlStore.ActionQueue {
+		queue = append(queue,
+			&fleetapi.ActionUpgrade{
+				ActionID:         a.ActionID,
+				ActionType:       a.Type,
+				ActionStartTime:  a.StartTime.Format(time.RFC3339),
```

Comment:
Does Format handle the case where the time is empty, in a different format, or completely malformed?